### ISSUE TITLE
Make frontend nginx proxy dynamically resolve DNS for data

### ISSUE THREAD

**Issue by scanon** [microbiomedata/nmdc-server#1051]
The nginx config has an entry to proxy the backend data service.  If the backend gets restarted or rescheduled on a new node, nginx will still have the IP cached and this will break downloads.

The fix is to change the nginx conf to use a variable name instead of data.  This will force it to resolve it on each requests.  I can give an example of what this would look like.

This is the relevant line....

https://github.com/microbiomedata/nmdc-server/blob/main/web/nginx.conf.template#L36

**Comment by eecavanna** [comment]
I found this Stack Overflow answer about using a variable to force re-resolution of DNS names:

https://serverfault.com/questions/240476/how-to-force-nginx-to-resolve-dns-of-a-dynamic-hostname-everytime-when-doing-p/593003#593003

**Comment by eecavanna** [comment]
In our case, usage of a variable would look like this:

```diff
          location /data/ {
+             set $data_proxy_host "data";
+             proxy_pass http://$data_proxy_host:8080/;
-             proxy_pass http://data:8080/;
              proxy_buffering off;
          }
```

There's a message on the [Nginx forum](https://forum.nginx.org/read.php?2,215830,215832#msg-215832) (linked to in that SO answer) saying that, even when using variables, the result will still be cached; but only for 5 minutes. I haven't seen this described in the official Nginx docs yet.

**Comment by eecavanna** [comment]
Once we specify a `resolver`, I think we can go even further and customize how long we want it to cache IP addresses for. Here's what I think that would look like in our case (for a 1-second cache TTL):

```diff
          location /data/ {
+             resolver 8.8.8.8 valid=1s;
+             set $data_proxy_host "data";
+             proxy_pass http://$data_proxy_host:8080/;
-             proxy_pass http://data:8080/;
              proxy_buffering off;
          }
```
> I assumed `0s` would not be a valid duration, but it might be.

**Comment by eecavanna** [comment]
In our case, usage of a variable would look like this: (CC: @naglepuff)

> Note: Someone on the [NERSC Users Slack workspace](https://nerscusers.slack.com/archives/C01LPA84AGM/p1679358990634609?thread_ts=1679166094.488819&cid=C01LPA84AGM) pointed out that, when taking this approach, he had to specify the hostname in the following format:
> ```
> http://<workload>.<namespace>.svc.cluster.local:<port>
> ```
> Note: that doesn't include the cluster name (i.e. `production`). Maybe that's not necessary because it's being resolved by something within the same cluster. :shrug:
>
> In our case, that would be:
> ```
> http://data.nmdc.svc.cluster.local:8080
> ```
> So, that's what I'll use below.

```diff
          location /data/ {
+             # Use a variable in the `proxy_pass ` directive so nginx attempts to re-resolve
+             # the domain to an IP address over time; instead of doing what it does by default,
+             # which is cache the IP address indefinitely after resolving it for the first time.
+             set $data_proxy_host "data.nmdc.svc.cluster.local";
+             proxy_pass http://$data_proxy_host:8080/;
-             proxy_pass http://data:8080/;
              proxy_buffering off;
          }
```

That SO answer says a "resolver" (i.e. name server) must be configured also. 


Whatever resolver we identify would have to be able to resolve the hostname (i.e. `data` or `data.nmdc.svc.cluster.local`... TBD) to the IP address of the data proxy workload (which I think rules out an external DNS server like Google's public DNS servers).

> The same person on NERSC Users Slack workspace said that he used the following resolver:
> ```
> 10.43.0.10
> ```
> I don't know what server that is. I assume it's a Spin-hosted DNS server that has a static IP address. The comments I will add to the blurbs below will reflect that assumption.

If we were to add a `resolver` directive with that IP address, the full config blurb would become:

```diff
          location /data/ {
+             # Use a Spin-hosted DNS server when resolving the hostname to an IP address.
+             resolver 10.43.0.10;
+
+             # Use a variable in the `proxy_pass ` directive so nginx attempts to re-resolve
+             # the domain to an IP address over time; instead of doing what it does by default,
+             # which is cache the IP address indefinitely after resolving it for the first time.
+             set $data_proxy_host "data";
+             proxy_pass http://$data_proxy_host:8080/;
-             proxy_pass http://data:8080/;
              proxy_buffering off;
          }
```

Once we specify a `resolver`, I think we can go even further and customize how long we want it to cache IP addresses for (I think it caches IP addresses for 5 minutes by default). Here's what that would look like in our case (for a 1-second cache lifetime):

```diff
          location /data/ {
+             # Use a Spin-hosted DNS server when resolving the hostname to an IP address,
+             # and only cache the IP address for 1 second.
+             resolver 10.43.0.10 valid=1s;
+
+             # Use a variable in the `proxy_pass ` directive so nginx attempts to re-resolve
+             # the domain to an IP address over time; instead of doing what it does by default,
+             # which is cache the IP address indefinitely after resolving it for the first time.
+             set $data_proxy_host "data";
+             proxy_pass http://$data_proxy_host:8080/;
-             proxy_pass http://data:8080/;
              proxy_buffering off;
          }
```

### References
- [The aforementioned StackOverflow answer](https://serverfault.com/questions/240476/how-to-force-nginx-to-resolve-dns-of-a-dynamic-hostname-everytime-when-doing-p/593003#593003)
- [Official docs about `set`](https://nginx.org/en/docs/http/ngx_http_rewrite_module.html#set)
- [Official docs about `proxy_pass`](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_pass)
- [Official docs about `resolver`](https://nginx.org/en/docs/http/ngx_http_core_module.html#resolver)
- [List of Google's public DNS servers](https://developers.google.com/speed/public-dns)
- [The aforementioned message on NERSC Users Slack workspace](https://nerscusers.slack.com/archives/C01LPA84AGM/p1679358990634609?thread_ts=1679166094.488819&cid=C01LPA84AGM)

**Comment by eecavanna** [comment]
Currently planning to fix by moving this functionality to the Ingress level. See https://github.com/microbiomedata/nmdc-server/issues/1094 for details and updates.

**Comment by ssarrafan** [comment]
@eecavanna can this issue be closed? Is this a duplicate for #1094?

**Comment by eecavanna** [comment]
Yes, this one can be closed. Yes, it is a duplicate in the sense that it stemmed from the same underlying issue. We have solved the underlying issue, although via a different means than was proposed in this ticket. In this ticket, we talked about reconfiguring Nginx. In reality, we ended up moving the path-based routing from Nginx to the Kubernetes Ingress. I will close this issue.
